### PR TITLE
Update enqueue functionality

### PR DESCRIPTION
**Pull request by dingo-d**
I've added an abstract class to avoid repeating the method calls for dependencies. Also, config interface has static methods so the handle names and other config related methods had to be changed into static calls.

This PR should enable people to add additional handlers in their enqueue classes with something like:

```php
use Eightshift_Libs\Enqueue\Enqueue_Admin;

class My_Enqeueue extends Enqueue_Admin {
  public function enqueue_scripts() {
    parent::enqueue_scripts();

    // Enqueue additional scripts?
  }

  protected function get_admin_script_dependencies() {
    return [ 'wp-pointer', 'jquery-ui-sortable' ];
  }

  protected function get_localizations() {
    return [
      'randomStuff' => esc_html__( 'Hi!', 'text-domain' );
    ];
  }
}
```

Feel free to test before approving this 😄

### CODE REVIEW

File: src/enqueue/class-enqueue-blocks.php [hunk 1]
```diff
@@ -108,60 +98,97 @@ public function enqueue_block_editor_script() {
...
+    $handler = "{$this->config::get_project_prefix()}-block-style";
 
     \wp_register_style(
       $handler,
       $this->manifest->get_assets_manifest_item( 'applicationBlocks.css' ),
-      [],
-      $this->config->get_project_version(),
+      $this->get_frontend_style_dependencies(),
+      $this->config::get_project_version(),
       false
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-blocks.php]: media override, could you please add it here also

File: src/enqueue/class-enqueue-admin.php [hunk 2]
```diff
@@ -73,17 +72,21 @@ public function register() {
...
+   *              Fixed static calls from config class.
    */
   public function enqueue_styles() {
-    $handler = "{$this->config->get_project_prefix()}-styles";
+    $handle = "{$this->config::get_project_prefix()}-styles";
 
     \wp_register_style(
-      $handler,
+      $handle,
       $this->manifest->get_assets_manifest_item( 'applicationAdmin.css' ),
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-admin.php]: we should think about method to change the name of the file?

File: src/enqueue/class-enqueue-admin.php [hunk 3]
```diff
@@ -93,17 +96,24 @@ public function enqueue_styles() {
...
+   *              Fixed static calls from config class.
    */
   public function enqueue_scripts() {
-    $handler = "{$this->config->get_project_prefix()}-scripts";
+    $handle = "{$this->config::get_project_prefix()}-scripts";
 
     \wp_register_script(
-      $handler,
+      $handle,
       $this->manifest->get_assets_manifest_item( 'applicationAdmin.js' ),
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-admin.php]: just like 🆙

File: src/enqueue/class-enqueue-blocks.php [hunk 4]
```diff
@@ -76,29 +76,19 @@ public function register() {
...
    *
    * @return void
    */
   public function enqueue_block_editor_script() {
-    $handler = "{$this->config->get_project_prefix()}-block-editor-scripts";
+    $handler = "{$this->config::get_project_prefix()}-block-editor-scripts";
 
     \wp_register_script(
       $handler,
       $this->manifest->get_assets_manifest_item( 'applicationBlocksEditor.js' ),
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-blocks.php]: just like 🆙

File: src/enqueue/class-enqueue-blocks.php [hunk 5]
```diff
@@ -108,60 +98,97 @@ public function enqueue_block_editor_script() {
...
    * @return void
    */
   public function enqueue_block_editor_style() {
-    $handler    = "{$this->config->get_project_prefix()}-block-editor-style";
-    $dependency = "{$this->config->get_project_prefix()}-block-style";
+    $handler = "{$this->config::get_project_prefix()}-block-editor-style";
 
     \wp_register_style(
       $handler,
       $this->manifest->get_assets_manifest_item( 'applicationBlocksEditor.css' ),
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-blocks.php]: just like 🆙

File: src/enqueue/class-enqueue-blocks.php [hunk 6]
```diff
@@ -108,60 +98,97 @@ public function enqueue_block_editor_script() {
...
    *
    * @return void
    */
   public function enqueue_block_style() {
-    $handler = "{$this->config->get_project_prefix()}-block-style";
+    $handler = "{$this->config::get_project_prefix()}-block-style";
 
     \wp_register_style(
       $handler,
       $this->manifest->get_assets_manifest_item( 'applicationBlocks.css' ),
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-blocks.php]: just like 🆙

File: src/enqueue/class-enqueue-blocks.php [hunk 7]
```diff
@@ -108,60 +98,97 @@ public function enqueue_block_editor_script() {
...
    *
    * @return void
    */
   public function enqueue_block_script() {
-    $handler = "{$this->config->get_project_prefix()}-block-scripts";
+    $handler = "{$this->config::get_project_prefix()}-block-scripts";
 
     \wp_register_script(
       $handler,
       $this->manifest->get_assets_manifest_item( 'applicationBlocks.js' ),
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-blocks.php]: just like 🆙

File: src/enqueue/class-enqueue-theme.php [hunk 8]
```diff
@@ -64,50 +64,52 @@ public function register() {
...
+   *              Fixed static calls from config class.
    */
   public function enqueue_styles() {
-    $handler = "{$this->config->get_project_prefix()}-theme-styles";
+    $handle = "{$this->config::get_project_prefix()}-theme-styles";
 
     \wp_register_style(
-      $handler,
+      $handle,
       $this->manifest->get_assets_manifest_item( 'application.css' ),
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-theme.php]: just like 🆙

File: src/enqueue/class-enqueue-theme.php [hunk 9]
```diff
@@ -64,50 +64,52 @@ public function register() {
...
 
     \wp_register_script(
-      $handler,
+      $handle,
       $this->manifest->get_assets_manifest_item( 'application.js' ),
-      [],
-      $this->config->get_project_version(),
+      $this->get_frontend_script_dependencies(),
+      $this->config::get_project_version(),
       true
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-theme.php]: just like 🆙

File: src/enqueue/class-enqueue-theme.php [hunk 10]
```diff
@@ -64,50 +64,52 @@ public function register() {
...
+   *              Fixed static calls from config class.
    */
   public function enqueue_scripts() {
-    $handler = "{$this->config->get_project_prefix()}-scripts";
+    $handle = "{$this->config::get_project_prefix()}-scripts";
 
     \wp_register_script(
-      $handler,
+      $handle,
       $this->manifest->get_assets_manifest_item( 'application.js' ),
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-theme.php]: just like 🆙

File: src/enqueue/class-enqueue-admin.php [hunk 11]
```diff
@@ -93,17 +96,24 @@ public function enqueue_styles() {
...
-      $this->config->get_project_version(),
+      $this->get_admin_script_dependencies(),
+      $this->config::get_project_version(),
       true
     );
-    \wp_enqueue_script( $handler );
+
+    \wp_enqueue_script( $handle );
+
+    foreach ( $this->get_localizations() as $object_name => $data_array ) {
```
Comment (iruzevic) [inline src/enqueue/class-enqueue-admin.php]: If there get_localizations is empty we can skip this loop. Can we add a conditional before it?

File: src/enqueue/class-enqueue-admin.php [hunk 12]
```diff
@@ -93,17 +96,24 @@ public function enqueue_styles() {
...
 
     \wp_register_script(
-      $handler,
+      $handle,
       $this->manifest->get_assets_manifest_item( 'applicationAdmin.js' ),
-      [],
-      $this->config->get_project_version(),
+      $this->get_admin_script_dependencies(),
+      $this->config::get_project_version(),
       true
```
[the remark this answers is not in the document]
Comment (dingo-d) [reply src/enqueue/class-enqueue-admin.php]: Thought about it, but then I thought that there was a good reason why you put it in the footer so I didn't change it 😄 I can add it :+1:

File: src/enqueue/class-enqueue-admin.php (hunk 2 above)
Comment (dingo-d) [reply src/enqueue/class-enqueue-admin.php]: Change the name of `applicationAdmin.css`? Isn't this webpack specific?

Comment (iruzevic) [reply src/enqueue/class-enqueue-admin.php]: yes but you can change it, if you change it in webpack we need to change it here also.

File: src/enqueue/class-enqueue-admin.php (hunk 11 above)
Comment (dingo-d) [reply src/enqueue/class-enqueue-admin.php]: No need. By default, the `get_localizations()` method returns an empty array, and if you pass am empty array to `foreach` nothing will happen (no error no nothing). So it's ok to leave it here 🙂

**Comment by iruzevic** [comment]
@dingo-d when do you think you will have this done?

**Comment by dingo-d** [comment]
Updated

**Comment by iruzevic** [comment]
Please add change log

**Comment by dingo-d** [comment]
@IvanGrginovInf @kmaric69 @gabriel-glo Please check this PR 🙂

**Comment by dingo-d** [comment]
I've updated the docblock @IvanGrginovInf

**Pull request merged**